File: 23-Graph/python/04_detect_a_cycle_in_an_undirected_graph_using_bfs.py
from collections import deque
from typing import List

class Solution:
    def isCycle(self, V: int, adj: List[List[int]]) -> bool:
        visited = [False] * V

        def bfs_check(src: int) -> bool:
            queue = deque([(src, -1)])
            visited[src] = True

            while queue:
                node, parent = queue.popleft()
                for neighbor in adj[node]:
                    if not visited[neighbor]:
                        visited[neighbor] = True
                        queue.append((neighbor, node))
                    elif neighbor != parent:
                        return True
            return False

        for i in range(V):
            if not visited[i]:
                if bfs_check(i):
                    return True
        return False

File: 23-Graph/cpp/04-detect-a-cycle-in-an-undirected-graph-using-bfs.cpp
#include <vector>
#include <queue>

using namespace std;

class Solution {
private:
    bool bfsCheck(int src, int V, vector<int> adj[], vector<bool>& visited) {
        queue<pair<int, int>> q;
        visited[src] = true;
        q.push({src, -1});

        while (!q.empty()) {
            auto [node, parent] = q.front();
            q.pop();

            for (int neighbor : adj[node]) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    q.push({neighbor, node});
                } else if (neighbor != parent) {
                    return true;
                }
            }
        }
        return false;
    }

public:
    bool isCycle(int V, vector<int> adj[]) {
        vector<bool> visited(V, false);
        for (int i = 0; i < V; i++) {
            if (!visited[i]) {
                if (bfsCheck(i, V, adj, visited)) {
                    return true;
                }
            }
        }
        return false;
    }
};
